### DISCUSSION TITLE
Implicit conversion between Fraction or BigNumber

### DISCUSSION THREAD

**Discussion by chrisbansart** [josdejong/mathjs#2587] [Q&A]
Hi,
I try to extend arithmetic operations with custom behaviors. Thanks to typed functions, I can set a default behavior for types operations. For instance, in the code below, I've chosen to convert into Fraction all operations that mix Fractions and bignumbers.

That works great and it prevents "exception : Cannot implicitly convert a Fraction to BigNumber or vice versa. Use function bignumber(x) to convert to BigNumber or fraction(x) to convert to Fraction."

Thus :
math.multiply(math.fraction(1/2),math.bignumber(3.0)) // gives Object { s: 1, n: 3, d: 2 }  which is a Fraction

My code 
```
const add = math.typed("add", {
  "Fraction, BigNumber": function(a, b) {
    return math.add(a, math.fraction(b));
  },

  "BigNumber, Fraction": function(a, b) {
    return math.add(math.fraction(a), b);
  },

  "Polynom, Polynom": function(a, b) {
    return Element_Algebra_Polynom.add(a, b);
  },
});

const subtract = math.typed("subtract", {
  "Fraction, BigNumber": function(a, b) {
    return math.subtract(a, math.fraction(b));
  },

  "BigNumber, Fraction": function(a, b) {
    return math.subtract(math.fraction(a), b);
  },
  "Polynom, Polynom": function(a, b) {
    return Element_Algebra_Polynom.subtract(a, b);
  },
});

const multiply = math.typed("multiply", {
  "Fraction, BigNumber": function(a, b) {
    return math.multiply(a, math.fraction(b));
  },

  "BigNumber, Fraction": function(a, b) {
    return math.multiply(math.fraction(a), b);
  },
});

const divide = math.typed("divide", {
  "Fraction, BigNumber": function(a, b) {
    return math.divide(a, math.fraction(b));
  },

  "BigNumber, Fraction": function(a, b) {
    return math.divide(math.fraction(a), b);
  },
});

math.import({
  add: add,
  substract: subtract,
  multiply: multiply,
  divide: divide,
});
```

Nevertheless I've faced more complicated cases where I'd like to apply this kind of conversions btw  a Matrix containing fraction or bignumber with a scalar fraction or bignumber
In this case : 
math.multiply([math.fraction(1,2),5],math.bignumber(1))  // raises error "TypeError: Cannot implicitly convert a Fraction to BigNumber or vice versa. Use function bignumber(x) to convert to BigNumber or fraction(x) to convert to Fraction."
But math.multiply([math.fraction(1,2),2],1)  works well

I imagine that my new type definitions don't apply element-wise in a matrix? but why ?

Any help would be greatly appreciated

Chris

**Comment by josdejong** [comment]
It could be related to the matrix implementations using separate functions for element wise operations: `addScalar`, `multiplyScalar`, `divideScalar`, `equalScalar`. The issue may be solved if you override those functions too.

**Reply by chrisbansart** [reply]
I've added
```
const addScalar = math.typed("addScalar", {
  "Fraction, BigNumber": function(a, b) {
    console.log("MyaddScalar");
    return a.add(math.fraction(b));
  },

  "BigNumber, Fraction": function(a, b) {
    console.log("MyaddScalar");

    return math.fraction(a).add(b);
  },
});

```
It works when I use directly the addScalar function, for instance `math.addScalar(math.bignumber(1),math.fraction(1,2))` 
but not with `math.add([math.fraction(1,2),2],math.bignumber(1))`. I guess that I miss something to extend addScalar within Matrix but don't see how to achieve.
Thanks
Chris

**Reply by josdejong** [reply]
Maybe `math.add` is still using the old `math.addScalar`? It is important to first create and import the updated `math.addScalar` and after that re-create `math.add` via a factory function where it will get the updated `addScalar` dependency. 

Maybe this example can be of help: https://mathjs.org/examples/advanced/custom_datatype.js.html
